### ISSUE TITLE
[Security] Strictly pin dependency versions

### ISSUE THREAD

**Issue by shannonwells** [LibertyDSNP/frequency#258]
Once we have settled on the versions we want for various dependencies we need to pin the versions.  
- [x] Pin Substrate version to whatever it is using #358 
- [x] Pin versions of this and other dependencies in Cargo.toml
- [x] Ensure that all Crates.io dependencies are using a specific version number
- [x] Examine dependabot 
- Update crates.io dependencies to be as up to date as possible (Separate PRs)
  - [x] Update non build breaking changes #403

**Comment by saraswatpuneet** [comment]
@wilwade  learning from other projects its ideal we keep main updated with latest version, but branch out before doing so and tag the parity version that branch is supposed to hold, say 0.9.25 which is the current one
What you think ?

Cumulus for example does that 

![image](https://user-images.githubusercontent.com/61435908/184876896-0c659874-a722-4a68-a37d-ab50ca16d003.png)

**Comment by wilwade** [comment]
@saraswatpuneet Hmm... Good question. We should do branches for versions. (#282 should document that, I'll add a note to it)

I don't know that we want pinning to be a required part of the release process however.

What about this setup?
- pin on main
- branch for releases
- commits to main to update pins

The goal being that we have explicit points where our dependencies change for dependency auditing purposes.

**Comment by johnmela** [comment]
I examined the pros / cons of pinning Git hosted dependencies with release branches vs commit hashes.

Cargo will allow either a branch pin:
`sp-std = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.27" }`
OR a rev pin
`sp-std = { git = "https://github.com/paritytech/substrate", default-features = false, rev = "8eff668a42325aeb4433eace1604f4d286a6ec05" }`
But NOT both.
We can either keep branch pinning to v0.9.27 and trust parity with updates / bug fixes / etc not breaking builds OR we can specify a commit hash and just comment which substrate / polkadot versions we are using.

After spinning my wheels on this some I've realized a caveat with the commit hash approach. We would have to lock each intra dependency that lives in the same GH repository to the same hash as the dependency that depends on it. Alcala patches each one in the rool Cargo.toml. For example,
frame_support would be locked to https://github.com/paritytech/substrate?branch=polkadot-v0.9.27#8eff668a42325aeb4433eace1604f4d286a6ec05" but it depends on sp-io which parity defines as `https://github.com/paritytech/substrate?branch=polkadot-v0.9.27
They both live in the same repository but the intra dependency (sp-io) will be set to the latest commit hash in the based on branch polkadot-v0.9.27, while the frame_support lib will be locked to an earlier state, at commit hash 8eff668a42325aeb4433eace1604f4d286a6ec05
This feels counter intuitive to how a package manager should work.
This approach would involve the up front work of locking each intra dependency to the same hash. Any new dependencies that live in substrate would have to be patched when we do version updates.
(This may be a reason moonbeam chose the maintain a fork route)
Perhaps we stay with the branch specification and let Cargo.lock do its job? The lock file locks each dependency and its intra dependencies to a hash revision. The only way we'd lose that hash lock is if we run a Cargo update or delete the lock file.
